utils/create_dataset.py

```diff
@@ -104,3 +114,155 @@ def __getitem__(self, index):
                                 f"\nIgnore if some augmentations have padded with dontcare value.")
         sample['index'] = index
         return sample
+
+
+class VRTDataset(GeoDataset):
+    def __init__(self, vrt_ds: DatasetReader) -> None:
+        """Initialize a new VRTDataset instance.
+        The dataset is base on the DataReader class, initiated by rasterio.open().
+
+        Args:
+            vrt_ds: DatasetReader object (rasterio)
+        """
+        super().__init__()
+
+        self.vrt_ds = vrt_ds
+        try:
+            self.cmap = vrt_ds.colormap(1)
+        except ValueError:
+            pass
+
+        crs = vrt_ds.crs
+        res = vrt_ds.res[0]
+
+        with WarpedVRT(vrt_ds, crs=crs) as vrt:
+            minx, miny, maxx, maxy = vrt.bounds
+
+        mint: float = 0
+        maxt: float = sys.maxsize
+
+        coords = (minx, maxx, miny, maxy, mint, maxt)
+        self.index.insert(0, coords, 'vrt')
+
+        self._crs = cast(CRS, crs)
+        self.res = cast(float, res)
+
+    def __getitem__(self, query: BoundingBox) -> Dict[str, Any]:
+        """Retrieve image and metadata indexed by query.
+
+        Args:
+            query: (minx, maxx, miny, maxy, mint, maxt) coordinates to index
+
+        Returns:
+            sample of image and metadata at that index
+        """
+        data = self._get_tensor(query)
+        key = "image"
+        sample = {key: data, "crs": self.crs, "bbox": query}
+
+        return sample
+
+    def _get_tensor(self, query):
+        """
+        Get a patch based on the given query (bounding box).
+        Args:
+            query:
+
+        Returns: Torch tensor patch.
+
+        """
+        bounds = (query.minx, query.miny, query.maxx, query.maxy)
+        out_width = round((query.maxx - query.minx) / self.res)
+        out_height = round((query.maxy - query.miny) / self.res)
+        out_shape = (self.vrt_ds.count, out_height, out_width)
+
+        dest = self.vrt_ds.read(
+            out_shape=out_shape, window=from_bounds(*bounds, self.vrt_ds.transform)
+        )
+
+        if dest.dtype == np.uint16:
+            dest = dest.astype(np.int32)
+        elif dest.dtype == np.uint32:
+            dest = dest.astype(np.int64)
+
+        tensor = torch.tensor(dest)
+
+        return tensor
+
+
+class GDLVectorDataset(GeoDataset):
+    """Abstract base class for :class:`GeoDataset` stored as vector files."""
+    def __init__(self, vec_ds: str = None, res: float = 0.0001) -> None:
+        """Initialize a new Dataset instance.
+
+        Args:
+            vec_ds: vector labels geopackage
+            res: resolution of the dataset in units of CRS
+        Returns:
+            An OGR datasource in memory
+        """
+        super().__init__()
+
+        self.vec_ds = ogr.Open(str(vec_ds))
+        self.res = res
+
+        # Populate the dataset index
+        try:
+            with fiona.open(str(vec_ds)) as src:
```

Comment:
I've been using geopandas in most places I deal with vector data in GDL (aoi, geoutils, tiling, etc.). Could we use geopandas here and below (see comment below)

Reply:
I used the OGR's method for fetching the vector file extents.